src/vivarium_public_health/metrics/mortality.py

```diff
@@ -71,99 +67,53 @@ def name(self):
-        ]
-
-    def on_collect_metrics(self, event: Event) -> None:
-        pop = self.population_view.get(event.index)
-        pop_died = pop[(pop["alive"] == "dead") & (pop["exit_time"] == event.time)]
-
-        groups = self.stratifier.group(
-            pop_died.index, self.config.include, self.config.exclude
-        )
-        for label, group_mask in groups:
-            for cause in self.causes_of_death:
-                cause_mask = pop_died["cause_of_death"] == cause
-                pop_died_of_cause = pop_died[group_mask & cause_mask]
-                new_observations = {
-                    f"death_due_to_{cause}_{label}": pop_died_of_cause.index.size,
-                    f"ylls_due_to_{cause}_{label}": pop_died_of_cause[
-                        "years_of_life_lost"
-                    ].sum(),
-                }
-                self.counts.update(new_observations)
-
-    ##################################
-    # Pipeline sources and modifiers #
-    ##################################
-
-    def metrics(self, index: pd.Index, metrics: Dict) -> Dict:
-        pop = self.population_view.get(index)
-
-        the_living = pop[(pop.alive == "alive") & pop.tracked]
-        the_dead = pop[pop.alive == "dead"]
-        metrics["years_of_life_lost"] = the_dead["years_of_life_lost"].sum()
-        metrics["total_population_living"] = len(the_living)
-        metrics["total_population_dead"] = len(the_dead)
-        metrics.update(self.counts)
-
-        return metrics
+        self.population_view = builder.population.get_view(columns_required)
+
+        for cause_of_death in self.causes_of_death:
+            builder.results.register_observation(
+                name=f"death_due_to_{cause_of_death}",
+                pop_filter=f'alive == "dead" and cause_of_death == "{cause_of_death}"',
+                aggregator=self.count_cause_specific_deaths,
+                requires_columns=["alive", "cause_of_death", "exit_time"],
+                additional_stratifications=self.config.include,
+                excluded_stratifications=self.config.exclude,
+                when="collect_metrics",
+            )
+            builder.results.register_observation(
+                name=f"ylls_due_to_{cause_of_death}",
+                pop_filter=f'alive == "dead" and cause_of_death == "{cause_of_death}"',
+                aggregator=self.calculate_cause_specific_ylls,
+                requires_columns=[
+                    "alive",
+                    "cause_of_death",
+                    "exit_time",
+                    "years_of_life_lost",
+                ],
+                additional_stratifications=self.config.include,
+                excluded_stratifications=self.config.exclude,
+                when="collect_metrics",
+            )
+
+    def count_cause_specific_deaths(self, x: pd.DataFrame) -> float:
+        died_of_cause = x["exit_time"] == self.clock() + self.step_size()
```

Comment:
Why not just `died_of_cause = x["exit_time"] > self.clock()`? Then you don't need to store `self.step_size`. The same applies below in YLLs